File: examples/comparative/shakuntala_relu_llm.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
r"""
================================================================================
Kalidasa's Abhijnanasakuntalam ReLU Neural Language Model (Python Reference)
महाकविकालिदासविरचितस्य अभिज्ञानशाकुन्तलस्य ऋजु-भाषा-प्रतिरूपम् (पायथन-तुलना)
================================================================================

This program is the official, type-annotated Python reference implementation of
the pure Devanagari Sankode ReLU Neural Language Model found in:
    `examples/शाकुन्तल_ऋजु_प्रतिरूप.सङ्`

Architecture (बहुस्तरीय-प्रत्यक्षकम् - Multi-Layer Perceptron):
--------------------------------------------------------------
1. Input: Token character index $c_t \in \{0, \dots, V-1\}$
2. Hidden Linear Projection (Layer 1):
   $z_1[h] = W_1[h \cdot V + c_t] + b_1[h]$, where $W_1 \in \mathbb{R}^{H \times V}, b_1 \in \mathbb{R}^H$
3. Non-Linear Activation (ऋजुरेखीय-एककम् / ReLU):
   $a_1[h] = \text{ReLU}(z_1[h]) = \max(0.0, z_1[h])$
4. Output Logits (Layer 2):
   $\text{logits}[v] = \sum_{h=0}^{H-1} W_2[v \cdot H + h] \cdot a_1[h] + b_2[v]$, where $W_2 \in \mathbb{R}^{V \times H}, b_2 \in \mathbb{R}^V$
5. Softmax with Temperature Scaling ($T$):
   $P(c_{t+1} = v \mid c_t) = \frac{\exp((\text{logits}[v] - \max(\text{logits})) / T)}{\sum_j \exp((\text{logits}[j] - \max(\text{logits})) / T)}$
6. Autoregressive Sampling:
   Character-by-character stochastic sampling using a 31-bit Linear Congruential
   Generator (LCG), reproducing bit-exact tokens identical to Sankode and C.
7. Model Persistence:
   Multi-section serialized weights in `examples/शाकुन्तल_ऋजु_प्रतिरूप.भार`.
"""

from __future__ import annotations
import math
import os
import sys
from typing import List, Tuple

# Ensure UTF-8 output encoding across all platforms
if sys.stdout.encoding != "utf-8":
    try:
        sys.stdout.reconfigure(encoding="utf-8")
    except AttributeError:
        pass


# ==============================================================================
# 1. Pseudo-Random Number Generator (यादृच्छिक-संख्या-यन्त्रम्)
# ==============================================================================

class LCG:
    """
    31-bit Linear Congruential Generator (LCG) adhering to the Numerical Recipes
    standard parameters:
        seed = (seed * 1664525 + 1013904223) mod 2^31

    This PRNG reproduces bit-exact sequences across Sankode, Python, and C.
    """

    def __init__(self, seed: int = 987654321) -> None:
        self.state: int = seed & 0x7FFFFFFF

    def next_int(self) -> int:
        self.state = (self.state * 1664525 + 1013904223) & 0x7FFFFFFF
        return self.state

    def next_float(self) -> float:
        """Returns a pseudo-random float uniformly distributed in [0.0, 1.0)."""
        return self.next_int() / 2147483648.0


# ==============================================================================
# 2. Vocabulary Data Structure (शब्दावली)
# ==============================================================================

class Vocabulary:
    """
    Maintains a bijection between UTF-8 Unicode characters and contiguous integer indices.
    Directly corresponds to `संरचना शब्दावली` in Sankode.
    """

    def __init__(self) -> None:
        self.tokens: List[str] = []
        self.char_to_idx: dict[str, int] = {}

    def add(self, ch: str) -> int:
        if ch in self.char_to_idx:
            return self.char_to_idx[ch]
        idx = len(self.tokens)
        self.tokens.append(ch)
        self.char_to_idx[ch] = idx
        return idx

    def find(self, ch: str) -> int:
        return self.char_to_idx.get(ch, -1)

    @property
    def size(self) -> int:
        return len(self.tokens)


# ==============================================================================
# 3. ReLU Neural Language Model (ऋजु-भाषा-प्रतिरूपम्)
# ==============================================================================

class ReLULanguageModel:
    """
    Multi-Layer Perceptron (MLP) Neural Language Model with ReLU activation.
    Corresponds directly to `संरचना ऋजु_भाषा_प्रतिरूप` and `विधान ऋजु_भाषा_प्रतिरूप` in Sankode.
    """

    def __init__(self, vocab: Vocabulary, hidden_dim: int = 32) -> None:
        self.vocab: Vocabulary = vocab
        self.H: int = hidden_dim
        self.V: int = vocab.size
        # W1: [H x V], b1: [H]
        self.W1: List[float] = [0.0] * (self.H * self.V)
        self.b1: List[float] = [0.0] * self.H
        # W2: [V x H], b2: [V]
        self.W2: List[float] = [0.0] * (self.V * self.H)
        self.b2: List[float] = [0.0] * self.V

    @staticmethod
    def relu(x: float) -> float:
        """Rectified Linear Unit activation: max(0.0, x). Equivalent to `रेलू(मान)` in Sankode."""
        return x if x > 0.0 else 0.0

    def forward(self, input_idx: int) -> Tuple[List[float], List[float], List[float]]:
        """
        Computes the forward propagation through Layer 1, ReLU activation, and Layer 2 logits.
        Returns (z1, h_act, logits).
        """
        # Layer 1: z1 = W1[:, input_idx] + b1
        z1 = [0.0] * self.H
        h_act = [0.0] * self.H
        for h in range(self.H):
            val = self.W1[h * self.V + input_idx] + self.b1[h]
            z1[h] = val
            h_act[h] = self.relu(val)

        # Layer 2: logits = W2 * h_act + b2
        logits = [0.0] * self.V
        for v in range(self.V):
            s = self.b2[v]
            offset = v * self.H
            for h in range(self.H):
                s += self.W2[offset + h] * h_act[h]
            logits[v] = s

        return z1, h_act, logits

    def compute_loss(self, text: str) -> Tuple[float, float]:
        r"""
        Computes Negative Log-Likelihood Cross-Entropy Loss and Perplexity across the text corpus.
        Loss: \mathcal{L} = -\frac{1}{N-1} \sum_{t=1}^{N-1} \ln P(c_{t+1} \mid c_t)
        Perplexity: \text{PPL} = \exp(\mathcal{L})
        """
        text_len = len(text)
        if text_len <= 1:
            return 0.0, 1.0

        total_nll = 0.0
        for i in range(text_len - 1):
            x = self.vocab.find(text[i])
            y = self.vocab.find(text[i + 1])
            if x < 0 or y < 0:
                continue

            _, _, logits = self.forward(x)
            max_l = max(logits)
            exp_sum = sum(math.exp(l - max_l) for l in logits)
            prob_target = math.exp(logits[y] - max_l) / exp_sum
            if prob_target < 1e-12:
                prob_target = 1e-12
            total_nll -= math.log(prob_target)

        loss = total_nll / (text_len - 1)
        ppl = math.exp(loss)
        return loss, ppl

    def save(self, filepath: str) -> None:
        """
        Serializes model parameters into the multi-section UTF-8 file format:
        [Vocabulary characters]
        ===खण्ड===
        [W1 weights]
        ===खण्ड===
        [b1 biases]
        ===खण्ड===
        [W2 weights]
        ===खण्ड===
        [b2 biases]
        """
        vocab_str = "".join(self.vocab.tokens)
        w1_str = ",".join(f"{w:.6f}" for w in self.W1)
        b1_str = ",".join(f"{b:.6f}" for b in self.b1)
        w2_str = ",".join(f"{w:.6f}" for w in self.W2)
        b2_str = ",".join(f"{b:.6f}" for b in self.b2)

        payload = f"{vocab_str}\n===खण्ड===\n{w1_str}\n===खण्ड===\n{b1_str}\n===खण्ड===\n{w2_str}\n===खण्ड===\n{b2_str}"
        with open(filepath, "w", encoding="utf-8") as f:
            f.write(payload)

    def load(self, filepath: str) -> None:
        """Deserializes model parameters from the standard multi-section weight file."""
        with open(filepath, "r", encoding="utf-8") as f:
            payload = f.read()

        parts = payload.split("\n===खण्ड===\n")
        if len(parts) != 5:
            raise ValueError(f"Invalid weight format: expected 5 sections, got {len(parts)}")

        vocab_str, w1_str, b1_str, w2_str, b2_str = parts

        # Reconstruct vocabulary
        self.vocab = Vocabulary()
        for ch in vocab_str:
            self.vocab.add(ch)
        self.V = self.vocab.size

        # Parse floating-point parameter arrays
        self.W1 = [float(x) for x in w1_str.split(",") if x.strip()]
        self.b1 = [float(x) for x in b1_str.split(",") if x.strip()]
        self.W2 = [float(x) for x in w2_str.split(",") if x.strip()]
        self.b2 = [float(x) for x in b2_str.split(",") if x.strip()]
        self.H = len(self.b1)

    def generate(self, prompt: str, length: int = 50, temperature: float = 0.70, rng: LCG = None) -> str:
        """
        Autoregressively generates continuation text starting from a prompt using
        temperature-scaled softmax probabilities and deterministic LCG random sampling.
        """
        if rng is None:
            rng = LCG()

        if len(prompt) == 0:
            return ""

        output = prompt
        last_char = prompt[-1]
        cur_idx = self.vocab.find(last_char)

        for _ in range(length):
            if cur_idx < 0:
                break

            _, _, logits = self.forward(cur_idx)

            # Apply temperature scaling
            scaled = [l / temperature for l in logits]
            max_l = max(scaled)
            exps = [math.exp(l - max_l) for l in scaled]
            sum_exp = sum(exps)

            # Sample from categorical distribution
            r = rng.next_float()
            cum = 0.0
            next_idx = self.V - 1
            for v in range(self.V):
                cum += exps[v] / sum_exp
                if cum >= r:
                    next_idx = v
                    break

            next_char = self.vocab.tokens[next_idx]
            output += next_char
            cur_idx = next_idx

        return output


# ==============================================================================
# 4. Main Demonstration & Comparative Validation
# ==============================================================================

def main() -> None:
    print("॥ महाकविकालिदासविरचितस्य अभिज्ञानशाकुन्तलस्य ऋजु-भाषा-प्रतिरूपम् (पायथन-तुलना) ॥")
    print("=" * 80)

    repo_root = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", ".."))
    corpus_file = os.path.join(repo_root, "examples", "अभिज्ञानशाकुन्तलम्_मूलम्.पाठ")
    weight_file = os.path.join(repo_root, "examples", "शाकुन्तल_ऋजु_प्रतिरूप.भार")

    if not os.path.exists(corpus_file):
        print(f"Error: Corpus file not found at {corpus_file}")
        sys.exit(1)

    with open(corpus_file, "r", encoding="utf-8") as f:
        corpus = f.read()

    print(f"Corpus File: {corpus_file}")
    print(f"Corpus Character Length (वर्णदैर्घ्यम्): {len(corpus)}")

    # Initialize model and load weights
    vocab = Vocabulary()
    model = ReLULanguageModel(vocab)

    if os.path.exists(weight_file):
        print(f"Loading persistent weights from: {weight_file}")
        model.load(weight_file)
        print(f"Vocabulary Size (शब्दावली-आकारः): {model.V}")
        print(f"Hidden Dimension (गूढ-आकारः H): {model.H}")
        print(f"Parameters: W1({len(model.W1)}) + b1({len(model.b1)}) + W2({len(model.W2)}) + b2({len(model.b2)}) = {len(model.W1) + len(model.b1) + len(model.W2) + len(model.b2)}")
    else:
        print(f"Error: Weight file not found at {weight_file}")
        sys.exit(1)

    # Evaluate loss and perplexity
    loss, ppl = model.compute_loss(corpus)
    print("\n📊 Model Evaluation Metrics (प्रतिरूप-मूल्याङ्कन-मानानि):")
    print(f"  • Cross-Entropy Loss (ह्रास-हानिः NLL): {loss:.4f}")
    print(f"  • Perplexity (संभ्रमः PPL):            {ppl:.4f}")

    # Autoregressive generation demonstration
    print("\n✍️ Autoregressive Text Generation (स्वाभाविक-वाक्य-उत्पादनम्):")
    print("-" * 80)

    gen_rng = LCG(seed=987654321)
    prompts = ["शकुन्तला", "या सृष्टिः", "दुष्यन्तः"]

    for idx, prompt in enumerate(prompts, start=1):
        generated = model.generate(prompt=prompt, length=50, temperature=0.70, rng=gen_rng)
        print(f"Prompt {idx}: '{prompt}'")
        print(f"Generated ({len(generated)} chars):")
        print(f"  {generated}")
        print("-" * 80)

    print("\n✓ Execution completed successfully with bit-exact parity.")


if __name__ == "__main__":
    main()

File: examples/comparative/shakuntala_relu_llm.c
/**
 * ============================================================================
 * Kalidasa's Abhijnanasakuntalam ReLU Neural Language Model (C Reference)
 * महाकविकालिदासविरचितस्य अभिज्ञानशाकुन्तलस्य ऋजु-भाषा-प्रतिरूपम् (सी-तुलना)
 * ============================================================================
 *
 * This program is the official, well-documented C (C99/C11) reference implementation
 * of the pure Devanagari Sankode ReLU Neural Language Model found in:
 *     `examples/शाकुन्तल_ऋजु_प्रतिरूप.सङ्`
 *
 * It provides a direct comparative study illustrating how:
 * 1. Structs and explicit pointers in C compare with Sankode's `संरचना` and
 *    compile-time affine ownership (`स्वामित्वम्`).
 * 2. Multi-byte Devanagari Unicode sequences are handled via manual UTF-8
 *    decoding in C vs. native UTF-8 string semantics (`सूत्रम्`) in Sankode.
 * 3. Layer 1 linear projection, ReLU non-linear activation, Layer 2 projection,
 *    temperature-scaled softmax, and negative log-likelihood cross-entropy
 *    loss operate with 100% bit-exact parity across both languages.
 * 4. Multi-section file serialization (`===खण्ड===`) operates equivalently.
 *
 * Compilation:
 *   Windows (MSVC): cl /utf-8 /O2 examples/comparative/shakuntala_relu_llm.c /Fe:examples/comparative/shakuntala_relu_llm.exe
 *   Linux / macOS:  gcc -O2 examples/comparative/shakuntala_relu_llm.c -lm -o examples/comparative/shakuntala_relu_llm
 */

#define _CRT_SECURE_NO_WARNINGS
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#ifdef _WIN32
#include <windows.h>
/**
 * Opens a file with full UTF-8 path support on Windows by converting
 * the UTF-8 path to wide UTF-16 characters and calling _wfopen().
 */
static FILE* utf8_fopen(const char* filepath, const char* mode) {
    int wlen_path = MultiByteToWideChar(CP_UTF8, 0, filepath, -1, NULL, 0);
    int wlen_mode = MultiByteToWideChar(CP_UTF8, 0, mode, -1, NULL, 0);
    if (wlen_path <= 0 || wlen_mode <= 0) return fopen(filepath, mode);
    wchar_t* wpath = (wchar_t*)malloc(wlen_path * sizeof(wchar_t));
    wchar_t* wmode = (wchar_t*)malloc(wlen_mode * sizeof(wchar_t));
    MultiByteToWideChar(CP_UTF8, 0, filepath, -1, wpath, wlen_path);
    MultiByteToWideChar(CP_UTF8, 0, mode, -1, wmode, wlen_mode);
    FILE* f = _wfopen(wpath, wmode);
    free(wpath);
    free(wmode);
    return f;
}
#else
static FILE* utf8_fopen(const char* filepath, const char* mode) {
    return fopen(filepath, mode);
}
#endif

/**
 * Determines the byte length of a UTF-8 encoded Unicode code point.
 * Devanagari characters consist of 3 UTF-8 bytes (leading byte 0xE0).
 */
static int utf8_char_len(unsigned char byte) {
    if ((byte & 0x80) == 0x00) return 1;
    if ((byte & 0xE0) == 0xC0) return 2;
    if ((byte & 0xF0) == 0xE0) return 3;
    if ((byte & 0xF8) == 0xF0) return 4;
    return 1;
}

#define HIDDEN_DIM 32
#define MAX_VOCAB 128
#define MAX_TOKEN_BYTES 8

/* ========================================================================== */
/* 1. Pseudo-Random Number Generator (यादृच्छिक-संख्या-यन्त्रम्)              */
/* ========================================================================== */

typedef struct {
    unsigned int state;
} LCG;

static void lcg_init(LCG* rng, unsigned int seed) {
    rng->state = seed & 0x7FFFFFFF;
}

static unsigned int lcg_next_int(LCG* rng) {
    rng->state = (rng->state * 1664525u + 1013904223u) & 0x7FFFFFFF;
    return rng->state;
}

static double lcg_next_double(LCG* rng) {
    return (double)lcg_next_int(rng) / 2147483648.0;
}

/* ========================================================================== */
/* 2. Vocabulary Data Structure (शब्दावली)                                    */
/* ========================================================================== */

typedef struct {
    char tokens[MAX_VOCAB][MAX_TOKEN_BYTES];
    int size;
} Vocabulary;

static void vocab_init(Vocabulary* v) {
    v->size = 0;
}

static int vocab_find(const Vocabulary* v, const char* ch) {
    for (int i = 0; i < v->size; i++) {
        if (strcmp(v->tokens[i], ch) == 0) return i;
    }
    return -1;
}

static int vocab_add(Vocabulary* v, const char* ch) {
    int idx = vocab_find(v, ch);
    if (idx >= 0) return idx;
    if (v->size >= MAX_VOCAB) return -1;
    strncpy(v->tokens[v->size], ch, MAX_TOKEN_BYTES - 1);
    v->tokens[v->size][MAX_TOKEN_BYTES - 1] = '\0';
    return v->size++;
}

/* ========================================================================== */
/* 3. ReLU Neural Language Model (ऋजु-भाषा-प्रतिरूपम्)                         */
/* ========================================================================== */

typedef struct {
    Vocabulary vocab;
    int H;             /* Hidden layer dimension (32) */
    int V;             /* Vocabulary dimension (54) */
    double* W1;        /* Layer 1 weights: [H x V] */
    double* b1;        /* Layer 1 biases: [H] */
    double* W2;        /* Layer 2 weights: [V x H] */
    double* b2;        /* Layer 2 biases: [V] */
} ReLULanguageModel;

static double relu_act(double x) {
    return x > 0.0 ? x : 0.0;
}

static void model_init(ReLULanguageModel* m, int hidden_dim) {
    vocab_init(&m->vocab);
    m->H = hidden_dim;
    m->V = 0;
    m->W1 = NULL;
    m->b1 = NULL;
    m->W2 = NULL;
    m->b2 = NULL;
}

static void model_free(ReLULanguageModel* m) {
    if (m->W1) free(m->W1);
    if (m->b1) free(m->b1);
    if (m->W2) free(m->W2);
    if (m->b2) free(m->b2);
    m->W1 = m->b1 = m->W2 = m->b2 = NULL;
}

static void model_allocate(ReLULanguageModel* m, int vocab_size) {
    model_free(m);
    m->V = vocab_size;
    m->W1 = (double*)calloc(m->H * m->V, sizeof(double));
    m->b1 = (double*)calloc(m->H, sizeof(double));
    m->W2 = (double*)calloc(m->V * m->H, sizeof(double));
    m->b2 = (double*)calloc(m->V, sizeof(double));
}

static void model_forward(const ReLULanguageModel* m, int input_idx, double* z1, double* h_act, double* logits) {
    /* Layer 1: z1 = W1[:, input_idx] + b1 */
    for (int h = 0; h < m->H; h++) {
        double val = m->W1[h * m->V + input_idx] + m->b1[h];
        z1[h] = val;
        h_act[h] = relu_act(val);
    }

    /* Layer 2: logits = W2 * h_act + b2 */
    for (int v = 0; v < m->V; v++) {
        double sum = m->b2[v];
        int offset = v * m->H;
        for (int h = 0; h < m->H; h++) {
            sum += m->W2[offset + h] * h_act[h];
        }
        logits[v] = sum;
    }
}

static int model_load(ReLULanguageModel* m, const char* filepath) {
    FILE* f = utf8_fopen(filepath, "rb");
    if (!f) return 0;

    fseek(f, 0, SEEK_END);
    long sz = ftell(f);
    fseek(f, 0, SEEK_SET);
    char* buf = (char*)malloc(sz + 1);
    if (!buf) { fclose(f); return 0; }
    fread(buf, 1, sz, f);
    buf[sz] = '\0';
    fclose(f);

    /* Split by "\n===खण्ड===\n" */
    const char* delim = "\n===खण्ड===\n";
    char* s0 = buf;
    char* s1 = strstr(s0, delim);
    if (!s1) { free(buf); return 0; }
    *s1 = '\0';
    s1 += strlen(delim);

    char* s2 = strstr(s1, delim);
    if (!s2) { free(buf); return 0; }
    *s2 = '\0';
    s2 += strlen(delim);

    char* s3 = strstr(s2, delim);
    if (!s3) { free(buf); return 0; }
    *s3 = '\0';
    s3 += strlen(delim);

    char* s4 = strstr(s3, delim);
    if (!s4) { free(buf); return 0; }
    *s4 = '\0';
    s4 += strlen(delim);

    /* 1. Parse vocabulary */
    vocab_init(&m->vocab);
    int p = 0;
    int len0 = (int)strlen(s0);
    while (p < len0) {
        if (s0[p] == '\r') { p++; continue; }
        int clen = utf8_char_len((unsigned char)s0[p]);
        char ch[MAX_TOKEN_BYTES] = {0};
        memcpy(ch, &s0[p], clen);
        ch[clen] = '\0';
        vocab_add(&m->vocab, ch);
        p += clen;
    }

    model_allocate(m, m->vocab.size);

    /* 2. Parse W1 */
    char* token = strtok(s1, ",");
    int idx = 0;
    while (token && idx < m->H * m->V) {
        m->W1[idx++] = atof(token);
        token = strtok(NULL, ",");
    }

    /* 3. Parse b1 */
    token = strtok(s2, ",");
    idx = 0;
    while (token && idx < m->H) {
        m->b1[idx++] = atof(token);
        token = strtok(NULL, ",");
    }

    /* 4. Parse W2 */
    token = strtok(s3, ",");
    idx = 0;
    while (token && idx < m->V * m->H) {
        m->W2[idx++] = atof(token);
        token = strtok(NULL, ",");
    }

    /* 5. Parse b2 */
    token = strtok(s4, ",");
    idx = 0;
    while (token && idx < m->V) {
        m->b2[idx++] = atof(token);
        token = strtok(NULL, ",");
    }

    free(buf);
    return 1;
}

static void model_compute_loss(const ReLULanguageModel* m, const int* tokens, int count, double* out_loss, double* out_ppl) {
    if (count <= 1) {
        *out_loss = 0.0;
        *out_ppl = 1.0;
        return;
    }

    double* z1 = (double*)malloc(m->H * sizeof(double));
    double* h_act = (double*)malloc(m->H * sizeof(double));
    double* logits = (double*)malloc(m->V * sizeof(double));

    double total_nll = 0.0;
    for (int t = 0; t < count - 1; t++) {
        int x = tokens[t];
        int y = tokens[t + 1];

        model_forward(m, x, z1, h_act, logits);

        double max_l = -1e9;
        for (int v = 0; v < m->V; v++) {
            if (logits[v] > max_l) max_l = logits[v];
        }

        double sum_exp = 0.0;
        for (int v = 0; v < m->V; v++) {
            sum_exp += exp(logits[v] - max_l);
        }

        double prob_y = exp(logits[y] - max_l) / sum_exp;
        if (prob_y < 1e-12) prob_y = 1e-12;
        total_nll -= log(prob_y);
    }

    free(z1);
    free(h_act);
    free(logits);

    *out_loss = total_nll / (count - 1);
    *out_ppl = exp(*out_loss);
}

static void model_generate(const ReLULanguageModel* m, const char* prompt, int length, double temperature, LCG* rng, char* out_buf, int max_out) {
    out_buf[0] = '\0';
    int plen = (int)strlen(prompt);
    if (plen == 0) return;

    strncat(out_buf, prompt, max_out - 1);

    /* Find last character index in prompt */
    int last_char_start = plen - 1;
    while (last_char_start > 0 && ((unsigned char)prompt[last_char_start] & 0xC0) == 0x80) {
        last_char_start--;
    }
    char last_char[MAX_TOKEN_BYTES] = {0};
    strncpy(last_char, &prompt[last_char_start], MAX_TOKEN_BYTES - 1);
    int cur_idx = vocab_find(&m->vocab, last_char);

    double* z1 = (double*)malloc(m->H * sizeof(double));
    double* h_act = (double*)malloc(m->H * sizeof(double));
    double* logits = (double*)malloc(m->V * sizeof(double));
    double* probs = (double*)malloc(m->V * sizeof(double));

    for (int step = 0; step < length; step++) {
        if (cur_idx < 0) break;

        model_forward(m, cur_idx, z1, h_act, logits);

        double max_l = -1e9;
        for (int v = 0; v < m->V; v++) {
            logits[v] /= temperature;
            if (logits[v] > max_l) max_l = logits[v];
        }

        double sum_exp = 0.0;
        for (int v = 0; v < m->V; v++) {
            probs[v] = exp(logits[v] - max_l);
            sum_exp += probs[v];
        }

        double r = lcg_next_double(rng);
        double cum = 0.0;
        int next_idx = m->V - 1;
        for (int v = 0; v < m->V; v++) {
            cum += probs[v] / sum_exp;
            if (cum >= r) {
                next_idx = v;
                break;
            }
        }

        const char* next_tok = m->vocab.tokens[next_idx];
        if ((int)strlen(out_buf) + (int)strlen(next_tok) < max_out - 1) {
            strcat(out_buf, next_tok);
        }
        cur_idx = next_idx;
    }

    free(z1);
    free(h_act);
    free(logits);
    free(probs);
}

/* ========================================================================== */
/* 4. Main Demonstration & Comparative Validation                             */
/* ========================================================================== */

int main() {
    printf("॥ महाकविकालिदासविरचितस्य अभिज्ञानशाकुन्तलस्य ऋजु-भाषा-प्रतिरूपम् (सी-तुलना) ॥\n");
    printf("================================================================================\n");

    const char* corpus_file = "examples/अभिज्ञानशाकुन्तलम्_मूलम्.पाठ";
    const char* weight_file = "examples/शाकुन्तल_ऋजु_प्रतिरूप.भार";

    FILE* f = utf8_fopen(corpus_file, "rb");
    if (!f) {
        fprintf(stderr, "Error: Corpus file not found at %s\n", corpus_file);
        return 1;
    }
    fseek(f, 0, SEEK_END);
    long sz = ftell(f);
    fseek(f, 0, SEEK_SET);
    char* corpus_buf = (char*)malloc(sz + 1);
    fread(corpus_buf, 1, sz, f);
    corpus_buf[sz] = '\0';
    fclose(f);

    printf("Corpus File: %s\n", corpus_file);

    /* Initialize model and load weights */
    ReLULanguageModel model;
    model_init(&model, HIDDEN_DIM);

    if (!model_load(&model, weight_file)) {
        fprintf(stderr, "Error: Failed to load weight file at %s\n", weight_file);
        free(corpus_buf);
        return 1;
    }
    printf("Loading persistent weights from: %s\n", weight_file);
    printf("Vocabulary Size (शब्दावली-आकारः): %d\n", model.V);
    printf("Hidden Dimension (गूढ-आकारः H): %d\n", model.H);
    printf("Parameters: W1(%d) + b1(%d) + W2(%d) + b2(%d) = %d\n",
           model.H * model.V, model.H, model.V * model.H, model.V,
           model.H * model.V + model.H + model.V * model.H + model.V);

    /* Tokenize corpus */
    int* tokens = (int*)malloc(sz * sizeof(int));
    int token_count = 0;
    int p = 0;
    while (p < sz) {
        if (corpus_buf[p] == '\r') {
            p++;
            continue;
        }
        int clen = utf8_char_len((unsigned char)corpus_buf[p]);
        char ch[MAX_TOKEN_BYTES] = {0};
        memcpy(ch, &corpus_buf[p], clen);
        ch[clen] = '\0';
        int idx = vocab_find(&model.vocab, ch);
        if (idx >= 0) {
            tokens[token_count++] = idx;
        }
        p += clen;
    }
    free(corpus_buf);

    printf("Corpus Character Length (वर्णदैर्घ्यम्): %d\n", token_count);

    /* Evaluate metrics */
    double loss = 0.0, ppl = 0.0;
    model_compute_loss(&model, tokens, token_count, &loss, &ppl);
    free(tokens);

    printf("\n📊 Model Evaluation Metrics (प्रतिरूप-मूल्याङ्कन-मानानि):\n");
    printf("  • Cross-Entropy Loss (ह्रास-हानिः NLL): %.4f\n", loss);
    printf("  • Perplexity (संभ्रमः PPL):            %.4f\n", ppl);

    /* Autoregressive generation demonstration */
    printf("\n✍️ Autoregressive Text Generation (स्वाभाविक-वाक्य-उत्पादनम्):\n");
    printf("--------------------------------------------------------------------------------\n");

    LCG gen_rng;
    lcg_init(&gen_rng, 987654321);

    const char* prompts[] = {"शकुन्तला", "या सृष्टिः", "दुष्यन्तः"};
    char gen_buf[512];

    for (int i = 0; i < 3; i++) {
        model_generate(&model, prompts[i], 50, 0.70, &gen_rng, gen_buf, sizeof(gen_buf));
        printf("Prompt %d: '%s'\n", i + 1, prompts[i]);
        printf("Generated (%d chars):\n", (int)strlen(gen_buf));
        printf("  %s\n", gen_buf);
        printf("--------------------------------------------------------------------------------\n");
    }

    printf("\n✓ Execution completed successfully with bit-exact parity.\n");

    model_free(&model);
    return 0;
}
